app/models/reservation.rb

```diff
@@ -17,4 +17,24 @@ class Reservation < ApplicationRecord # :nodoc:
   validates :checkin_at, presence: true
   validates :checkout_at, presence: true
   validates :book_code, presence: true, uniqueness: true
+  validate :room_must_be_available, if: -> { room_id.present? || checkin_at.present? || checkout_at.present? }
```

Comment:
I don't understand the if conditions. You are already validating `checkin_at` and `checkout_at` before, why do you have this in this validations again? Also putting them in OR conditions, means if there is no `room_id` and there is `checkin_at`, you will run this validation? I don't understand what is the use case for this logic.